File: ansible_aisnippet/providers/factory.py
"""Provider Factory – creates AI provider instances by name (Factory Pattern)."""
from __future__ import annotations

from typing import TYPE_CHECKING

from .base import BaseProvider

if TYPE_CHECKING:
    from ..config import ProviderConfig

_REGISTRY: dict[str, type[BaseProvider]] = {}


def _register(provider_class: type[BaseProvider]) -> type[BaseProvider]:
    """Register a provider class under its ``name`` attribute."""
    _REGISTRY[provider_class.name] = provider_class
    return provider_class


def _lazy_register_all() -> None:
    """Import and register all built-in providers (deferred to first use)."""
    if _REGISTRY:
        return

    from .anthropic_provider import AnthropicProvider
    from .azure_provider import AzureOpenAIProvider
    from .cohere_provider import CohereProvider
    from .google_provider import GoogleGeminiProvider
    from .huggingface_provider import HuggingFaceProvider
    from .llama_provider import MetaLlamaProvider
    from .lmstudio_provider import LMStudioProvider
    from .mistral_provider import MistralProvider
    from .ollama_provider import OllamaProvider
    from .openai_provider import OpenAIProvider
    from .openrouter_provider import OpenRouterProvider
    from .zen_provider import ZenProvider

    for cls in [
        OpenAIProvider,
        AnthropicProvider,
        GoogleGeminiProvider,
        AzureOpenAIProvider,
        MistralProvider,
        CohereProvider,
        OllamaProvider,
        LMStudioProvider,
        MetaLlamaProvider,
        HuggingFaceProvider,
        OpenRouterProvider,
        ZenProvider,
    ]:
        _register(cls)


class ProviderFactory:
    """Factory for creating AI provider instances.

    Usage::

        from ansible_aisnippet.providers.factory import ProviderFactory
        from ansible_aisnippet.config import ProviderConfig

        cfg = ProviderConfig(api_key="sk-...", model="gpt-4")
        provider = ProviderFactory.create("openai", cfg)
        response = provider.generate(system_msg, user_msg)
    """

    @staticmethod
    def create(provider_name: str, config: "ProviderConfig") -> BaseProvider:
        """Instantiate a provider by name.

        Args:
            provider_name: One of the supported provider identifiers
                (``openai``, ``anthropic``, ``google``, ``azure``,
                ``mistral``, ``cohere``, ``ollama``, ``lmstudio``,
                ``llama``, ``huggingface``, ``openrouter``, ``zen``).
            config: A :class:`~ansible_aisnippet.config.ProviderConfig`
                containing API key, model, and other settings.

        Returns:
            A :class:`BaseProvider` instance ready to call.

        Raises:
            ValueError: If *provider_name* is not registered.
        """
        _lazy_register_all()
        name_lower = provider_name.lower().strip()
        provider_class = _REGISTRY.get(name_lower)
        if provider_class is None:
            available = ", ".join(sorted(_REGISTRY.keys()))
            raise ValueError(
                f"Unknown provider '{provider_name}'. "
                f"Available providers: {available}"
            )
        return provider_class(config)

    @staticmethod
    def list_providers() -> list[str]:
        """Return a sorted list of all registered provider names."""
        _lazy_register_all()
        return sorted(_REGISTRY.keys())

File: internal/providers/factory.go
package providers

import (
	"fmt"
	"strings"

	"github.com/marcuscabrera/ansible-aisnippet/internal/config"
)

// registry maps provider name → constructor function.
var registry = map[string]func(config.ProviderConfig) Provider{}

// register adds a provider constructor to the registry.
func register(name string, ctor func(config.ProviderConfig) Provider) {
	registry[name] = ctor
}

func init() {
	register("openai", func(cfg config.ProviderConfig) Provider { return newOpenAI(cfg) })
	register("anthropic", func(cfg config.ProviderConfig) Provider { return newAnthropic(cfg) })
	register("google", func(cfg config.ProviderConfig) Provider { return newGoogle(cfg) })
	register("azure", func(cfg config.ProviderConfig) Provider { return newAzure(cfg) })
	register("mistral", func(cfg config.ProviderConfig) Provider { return newMistral(cfg) })
	register("cohere", func(cfg config.ProviderConfig) Provider { return newCohere(cfg) })
	register("ollama", func(cfg config.ProviderConfig) Provider { return newOllama(cfg) })
	register("lmstudio", func(cfg config.ProviderConfig) Provider { return newLMStudio(cfg) })
	register("llama", func(cfg config.ProviderConfig) Provider { return newLlama(cfg) })
	register("huggingface", func(cfg config.ProviderConfig) Provider { return newHuggingFace(cfg) })
	register("openrouter", func(cfg config.ProviderConfig) Provider { return newOpenRouter(cfg) })
	register("zen", func(cfg config.ProviderConfig) Provider { return newZen(cfg) })
}

// Create instantiates a provider by name with the given config.
//
// provider_name must be one of: openai, anthropic, google, azure, mistral,
// cohere, ollama, lmstudio, llama, huggingface, openrouter, zen.
//
// Returns an error if the provider name is not registered.
func Create(providerName string, cfg config.ProviderConfig) (Provider, error) {
	name := strings.ToLower(strings.TrimSpace(providerName))
	ctor, ok := registry[name]
	if !ok {
		available := ListProviders()
		return nil, fmt.Errorf("unknown provider %q; available providers: %s",
			providerName, strings.Join(available, ", "))
	}
	return ctor(cfg), nil
}

// ListProviders returns a sorted list of all registered provider names.
func ListProviders() []string {
	names := make([]string, 0, len(registry))
	for name := range registry {
		names = append(names, name)
	}
	// Simple sort
	for i := 0; i < len(names); i++ {
		for j := i + 1; j < len(names); j++ {
			if names[i] > names[j] {
				names[i], names[j] = names[j], names[i]
			}
		}
	}
	return names
}
